Don't let the UTF8OutPipeReader be static

DHowett-MSFT opened the pull request:
Whoops. This fixes #1997.

DHowett-MSFT commented:
/cc @german-one (i can't request a review from you, but you can still sign off if this makes sense :smile:)

german-one commented:
Thanks for keeping me updated!

How embarrassing for me. I discussed with @miniksa about a static class instance in `WriteConsoleA` which seems to made me blind for the different preconditions here. However, there is no excuse.

Would you mind to update this line too?
https://github.com/microsoft/terminal/blob/8d52ba0990f89de552b572381adb26b77cee847b/src/types/ut_types/UTF8OutPipeReaderTests.cpp#L124
`UTF8OutPipeReader reader{ outPipe }; // declare an instance of UTF8OutPipeReader`  
Not that this would make any difference here. Just to reconcile it with the productive code.

DHowett-MSFT commented:
Sure! Thanks for the catch! :smile:

german-one commented on src/types/UTF8OutPipeReader.cpp:
```diff
@@ -32,13 +32,13 @@ UTF8OutPipeReader::UTF8OutPipeReader(wil::unique_hfile& outPipe) :
+        return E_FAIL;
     }
-    else if (!fSuccess) // reading failed
+    else if (dwRead == 0) // quit if no data has been read and no cached data was left over
```
Once you update this, consider to make `else if` just an `if`. The `else` is completely meaningless here.

german-one commented:
Hold on a second
![image](https://user-images.githubusercontent.com/46659645/61394689-f9876580-a8c3-11e9-870d-770cc6517893.png)
Seems updating the Unit Test was a real good idea. I already have a rough idea why this fails. It's likely that we can't work with `std::string_view&` any longer in `UTF8OutPipeReader::Read`.

DHowett-MSFT commented:
Unless the individual calls to `RunTest` were intended to rely on static state stored in the UTF8OPutPipeReader (partials from one test flowing into the next test), making it non-static was correct.

german-one commented:
Yes this was definitely correct. I feel responsible for that. Just give me some minutes figuring out what's going wrong ...

german-one commented:
Having `if (!fSuccess)` before `if (dwRead == 0)` makes it fail. Even though I agree that this order is correct...

german-one commented:
@DHowett-MSFT Would that be reasonable?
```
    if (!fSuccess) // reading failed
    {
        if (GetLastError() == ERROR_BROKEN_PIPE) // if the pipe has been ended
        {
            return S_OK;
        }

        return E_FAIL;
    }

    if (dwRead == 0) // quit if no data has been read and no cached data was left over
    {
        return S_OK;
    }
```

And would that be redundant in this case?
https://github.com/microsoft/terminal/blob/8d52ba0990f89de552b572381adb26b77cee847b/src/cascadia/TerminalConnection/ConhostConnection.cpp#L202

german-one commented:
The reason why I originally checked `if (dwRead == 0)` first was because of that piece of the comment:

> // quit if no data has been read **and no cached data was left over**

Now that we escape if `RadFile` failed, we may still have data in the buffer that has been cached before.

DHowett-MSFT commented:
If there's data in the buffer that was cached, but ReadFile failed, that data can't be used. It's still a partial UTF-8 sequence!

german-one commented:
Absolutely. So we could just discard it.

DHowett-MSFT commented:
Okay, this is now a little different. I've cleaned up some of the code and added another _successful_ exit case (reading from a closed pipe is a detectable 0-byte "successful" read.)

DHowett-MSFT commented:
I switched away from using `unique_hfile` because that's only supposed to be used when a class _owns_ a handle. Holding a reference to a unique_hfile is the same as holding a raw HANDLE!

german-one commented:
I like this. Looks a little like my proposal above.

german-one commented on src/types/UTF8OutPipeReader.cpp:
```diff
@@ -6,19 +6,32 @@
+//   S_OK for a successful read
+//   S_FALSE for a read on a closed pipe
+//   E_* (anything) for a failed read
 [[nodiscard]] HRESULT UTF8OutPipeReader::Read(_Out_ std::string_view& strView)
```
Thanks for adding this!

german-one commented on src/types/UTF8OutPipeReader.cpp:
```diff
@@ -27,18 +40,29 @@ UTF8OutPipeReader::UTF8OutPipeReader(wil::unique_hfile& outPipe) :
     {
-        return S_OK;
+        auto lastError = GetLastError();
+        if (GetLastError() == ERROR_BROKEN_PIPE)
```
`if (lastError == ERROR_BROKEN_PIPE)` maybe?

DHowett-MSFT replied:
Whoops, good catch!
